### PR TITLE
add IR graph and conversion from IR graph to blas graph or dnn graph

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?
1. implement IR graph
2. convert Ir graph to blas graph
3. convert Ir graph to dnn graph, including: add input wrapper, enhancement DnnGraph, etc.

## How was this patch tested?
unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/InputWrapper.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn.mkldnn
+
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+
+private[bigdl] class InputWrapper extends MklDnnLayer {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/InputWrapper.scala]: why do we need  a separate wrapper for input ? we already have BLASWrapper

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import com.intel.analytics.bigdl.utils.{MklBlas, MklDnn, Node}
+
+import scala.reflect.ClassTag
+
+
+private[bigdl] class IRConverter[T: ClassTag](IRgraph: IRGraph[T])(implicit ev: TensorNumeric[T]) {
+  /**
+   * convert IRgraph to blas or dnn graph.
+   * @param dnnMode If dnnMode = true, it means ir graph must be converted to dnn graph
+   * @return
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: Add return desc ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  def toGraph(dnnMode: Boolean = false) : Graph[T] = {
+    if (utils.Engine.getEngineType() == MklBlas && !dnnMode) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: is underlying engine related to conversion ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      require(IRToBlas[T].convertingCheck(IRgraph.allNodes.toArray),
+        "IR graph can not convert to Blas layer")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: cannot be converted

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      toBlasGraph()
+    } else {
+      require(ev.getType() == FloatType, "Mkldnn engine only supports float data")
+      utils.Engine.setEngineType(MklDnn)
+      require(IRToDnn[Float].convertingCheck(
+        IRgraph.allNodes.toArray.asInstanceOf[Array[Node[IRElement[Float]]]]),
+        "IR graph can not convert to Dnn layer")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: cannot be converted

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala (hunk 3 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: I see you check the engine type when converting to BLAS, but you don't check when converting to DNN, instead you just   set the engine type to DNN for DNN conversion

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{Engine, MklBlas, Node, T}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Generate IR graph
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: Generate -> General ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+  private def initFwdPrimitives(input: Activity): Unit = {
+    if (!initFwd && graph.isInstanceOf[DnnGraph]) {
+      val inputMemory = new Array[MemoryData](inputFormats.length)
+      if (input.isInstanceOf[Tensor[T]]) {
+        inputMemory(0) = HeapData(input.toTensor[T].size(), inputFormats(0))
+      } else {
+        val tensors = input.toTable
+        require(tensors.length() == inputFormats.length, s"table input length " +
+          s"${tensors.length()} should be same with inputFormats length ${inputFormats.length}")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: same => the same

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        tensors.foreach(t => {
+          require(t._2.isInstanceOf[Tensor[T]],
+            "Only support input just contains tensor, but here get Table in input")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: Only support input with tensor type, table not supported

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/InputWrapper.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/InputWrapper.scala]: BLASWrapper is only for blas layer wrapper. And InputWrapper is only to wrap dnn input layer.
For dnn input layer, it needs  input shape (usually batch size and image size) and format to init.
So when we create a dnn model, we must have input shape and can not be changed.  It means inference must use the same batch size with training.

In order to fix this issue, I made two points change:
1. create primitives in IRgraph only when forward / backward is called.
2. Init dnn input layer in InputWrapper, not creating dnn model

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: ok, delete dnnMode and do some change

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRConverter.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala (hunk 7 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala (hunk 8 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala (hunk 6 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: I think it is not general, but generate a ir graph from ir elements

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: Got it

**Comment by zhangxiaoli73** [comment]
Jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/314/

**Pull request merged**